README.md
# Cpick

![1.2.0](https://img.shields.io/badge/status-1.2.0-red)
[![GoDoc](https://godoc.org/github.com/ethanbaker/cpick?status.svg)](https://godoc.org/github.com/ethanbaker/cpick)
[![Go Report Card](https://goreportcard.com/badge/github.com/ethanbaker/cpick)](https://goreportcard.com/report/github.com/ethanbaker/cpick)
[![Coverage](https://gocover.io/_badge/github.com/ethanbaker/cpick)](https://gocover.io/github.com/ethanbaker/cpick)


An extensive color picker in the terminal.

![color picker image](logo.gif)

---

## About

Cpick is an interactive color picker in the terminal. You can run Cpick in any true color terminal, and you can see thousands of unique colors, either from preset values or gradients. Each color has its own formats in many different forms, including as RGB, HSV, CMYK, and more. 

---

## Dependencies

Cpick depends on [Tcell](github.com/gdamore/tcell) and [Colors](github.com/ethanbaker/colors).

Cpick also uses [Cview](gitlab.com/tslocum/cview). However, Cview currently has a feature where the table cells are drawn with a required space in between cells, which ruins the aesthetic of Cpick. In order to fix this, a forked version of Cview is used within the Cpick package that has the required fixes. This may lead to bugs and issues in older versions of Cview that are not able to be readily fixed until Cview adopts required features.

---

## Installation

#### Golang 

To include the Cpick package, include the line `import "github.com/ethanbaker/cpick"`.

To make a command based off of the package, run `go install` in the `cmd/cpick` directory. Command usage can be found in the [docs](https://godoc.org/github.com/ethanbaker/cpick).

Cpick also utilizes [cmdtab](https://github.com/rwxrob/cmdtab), which offers tab completion for Cpick. If you wish to enable tab completion, add the line `complete -C cpick cpick` to your `~/.bashrc` file.

#### Arch Linux (Arch User Repository)

Visit the [AUR page](https://aur.archlinux.org) to download Cpick to your arch system. If you don't know how to download a package from the Arch User Repository, you can follow directions from the [Arch wiki](https://wiki.archlinux.org/index.php/Arch_User_Repository#Installing_packages).

#### Ubuntu/Debian

Coming soon!

---

## Documentation

[Documentation](https://godoc.org/github.com/ethanbaker/cpick) is present using Godoc.

---

## Features

#### Cpick Settings

Cpick comes with two boolean settings: sandbox and testing.

* Sandbox

Sandbox determines whether or not Cpick will return a color. Sandbox is useful for looking at colors that will be used somewhere else, like for a website. If sandbox is set to `true`, Cpick will not return any value but will instead cycle between different color tables. Setting sandbox to `false` means a color will be returned. 

* Testing

Testing is used to test coverage and validness of Cpick without requiring user input. Testing is useful for making sure Cpick can work on your device. Setting testing to `true` turns on the test mode, while setting testing to `false` keeps Cpick on normal mode.

All of the tests can be found in the [tests.go](https://github.com/ethanbaker/cpick/blob/master/tests.go) file.

#### Controls/Keys

If you are looking for a simplified version, check out the [docs](https://github.com/ethanbaker/cpick/blob/master/colors.go). Here is a more in-depth explanation.

You can also see a minified help version by pressing \` while the Cpick application is running.

In Cpick, there are two main screens: the **hue screen** and the **saturation-value screen**.

The **hue screen** is the first screen that pops up. It contains a hue slider present at the top, two color tint previews on the left, and a table of different preset or custom colors at the bottom.

The **saturation-value screen** is the screen that pops up once you press enter to select a color on the **hue screen**. It contains one large color gradient in the center and two color tint previews on the right.

On the **hue screen**, you initially start off on the hue slider. This is where you can select a specific hue by navigating the table, either using h and l (from vim) or the left and right arrow. If you press enter, you will be moved to the **saturation-value** screen with the table representing the hue you chose. You can also press space to move to the listed colors at the bottom of the **hue screen**.

For the listed color table on the **hue screen**, you can navigate it using the standard vim keys (h, j, k, l) or the arrow keys. If you press enter on a selected color, you will be taken to the **saturation-value screen** and the color you selected will automatically be selected on the big color gradient table. You can also switch between color types by pressing c to go forward a page and C to go back.

In addition, you can press ? to open up a search menu. Here you can search for a specific color or keyword. You can press N to go to the next selection and n to go to the previous selection, just like in vim. 

On the **saturation-value screen**, you can move about the screen using the standard vim keys (h, j, k, l) or the arrow keys. Once on a desired color, press enter to select it (take note that if you ran Cpick with sandbox as true, pressing enter on the table will take you back to the **hue screen**).

In addition, you can press g to go to the start of the table (top-left most cell) or G to go to the end of the table (bottom-right most cell).

#### Custom Colors

In Cpick, you can add custom colors that can come up on the color pages. You can add JSON files that hold the colors in 3 ways.

1. **Local Environment**

Wherever you are running Cpick, you can provide a local `colors.json` file (file would have the path `./colors.json` from wherever Cpick is being run). This has the highest priority.

2. **~/.config Directory**

In the `~/.config` directory, you can create a `cpick` directory that can contain the `colors.json` file (file would have the path `~/.config/cpick/colors.json`). This has a lower priority than a local file but a higher priority than a `./cpick` directory..

3. **~/.cpick Directory**

In your home directory, you can create a `.cpick` directory that can contain the `colors.json` file (file would have the path `~/.cpick/colors.json`). This has the lowest priority.

The `colors.json` file has a very strict format. 

~~~json
{
  "colorList": [
    {
      "name": "COLOR NAME 1",
      "colors": []
    },

    {	
      "name": "COLOR NAME 2",
      "colors": []
    },

    {	
      "name": "COLOR NAME 3",
      "colors": []
    }

  ]
}
~~~

Each color type consists of an object with a `name` and `colors` key. The `name` key consists of what color type the colors provided are, such as CSS, Solarized, or XTERM. The `colors` key holds all of the colors that will be previewed when Cpick is run. All of the different color types are in the `colorList` array.

An individual color is an object that consists of two keys: `name` and `value`. `name` is the name of the color and `value` is the hexadecimal value of the color as a string. The "#" for the hexadecimal value is optional.

~~~json
"colors": [
  {"name": "Red",   "value": "#FF0000"},
  {"name": "Green", "value": "#00FF00"},
  {"name": "Blue",  "value": "#0000FF"}
]
~~~

Cpick comes with three color types as a default: CSS, Solarized, and XTERM. In order to fix complicated import problems, the JSON data is present in the [colors.go](https://github.com/ethanbaker/cpick/blob/master/colors.go) file as a string. The preset data always has the lowest priority for being used.



cview/CHANGELOG
v1.5.3 (2020-01-14)
- Document how to prevent screen artifacts when using SetBackgroundTransparent
- Fix highlighting focused Form element
- Fix incorrect TabbedPanels colors
- Preserve order of panels when updating panel

v1.5.2 (2020-12-04)
- Handle input events before executing queued update functions
- Fix WindowManager background not being drawn

v1.5.1 (2020-11-05)
- Add FocusManager
- Add Slider
- Add TabbedPanels
- Add Application.Init
- Add Application.GetScreen and Application.GetScreenSize
- Add SetVisible and GetVisible to all widgets
- Add TextView.SetBytes, TextView.GetBytes and TextView.SetWrapWidth
- Add TableCell.SetBytes, TableCell.GetBytes and TableCell.GetText
- Fix List dividers allowing selection
- Fix List.Transform not calling handler set via SetChangedFunc
- Fix WordWrap bounds out of range
- Fix WordWrap dropping last character of closing tag
- Allow modification of scroll bar render text
- Allow scrolling List horizontally
- Clarify that Table rows must each have the same number of columns
- Clarify that WordWrap accepts unescaped text only
- Display TextView scroll bar automatically by default
- Generalize tag stripping as StripTags
- Make printWithStyle public and rename as PrintStyle
- Optimize TextView (writing is 90% faster, drawing is 50% faster)
- Reduce Box inner rect calculations
- Remove return values from methods which return their primitive (breaks chaining)
- Remove Application.ForceDraw (Application.Draw may be called anywhere)
- Rename SetBorderPadding and GetBorderPadding as SetPadding and GetPadding
- Rename Pages as Panels
- Support bracketed paste mode via tcell

v1.5.0 (2020-10-03)
- Add scroll bar to TextView
- Add Window and WindowManager
- Add focus-driven style options
- Add InputField autocomplete style options
- Add arrow symbol to DropDown
- Add Makefile
- Allow autocomplete selection value customization
- Provide DropDownOption in DropDown handlers
- Provide ListItem in List handlers
- Panic when attempting to add an invalid FormItem
- Release lock on InputField while executing DoneFunc and FinishedFunc

v1.4.9 (2020-09-08)
- Add InputField.GetCursorPosition and InputField.SetCursorPosition
- Add Table.Sort, Table.SetSortFunc and Table.SetSortClicked
- Add TextView.SetReindexBuffer
- Base List PageUp and PageDown on the number of items displayed
- Upgrade tcell to v2: includes strikethrough support

v1.4.8 (2020-08-11)
- Add italic text formatting flag
- Add Modal.GetForm and Modal.GetFrame
- Fix Form.Clear deadlock
- Fill nil Flex space with default background color
- Position ContextMenu on selected item when negative coordinates are provided
- Use sync.RWMutex in all widgets

v1.4.7 (2020-06-09)
- Add Box.SetBackgroundTransparent
- Add List.SetSelectedAlwaysCentered
- Fix default background transparency of Flex and Grid
- Fix negative List offset and DropDown selection
- Fix panic when clicking nil primitive in Flex
- Fix deadlock when calling Pages.SendToFront
- Fix ANSI colors 0-15
- Fix ANSI SGR codes

v1.4.6 (2020-05-18)
- Add Box.ShowFocus
- Add Keys to allow default keyboard shortcuts to be modified
- Add List.GetOffset, List.SetOffset and List.SetSelectedTextAttributes
- Add TextView.SetMaxLines
- Add Vim-style keybindings to List
- Fix List not updating selected item before calling selected handlers
- Do not handle right click on List when there is no context menu
- Always initialize context menu List
- Document how to override default keybindings

v1.4.5 (2020-04-25)
- Add multithreading support
- Add ContextMenu (initially supported by List)
- Add List.Transform and TreeView.Transform
- Add Application.SetBeforeFocusFunc and Application.SetAfterFocusFunc
- Merge upstream mouse support

v1.4.4 (2020-02-24)
- Fix panic when navigating empty list
- Fix resize event dimensions on Windows
- Clarify that Box does not have inner text
- Do not wrap around form by default

v1.4.3 (2020-02-13)
- Add SetFocusedFunc to TreeNode
- Add option to always show scroll bar
- Fix scrolling Table with PageDown and PageUp
- Do not wrap around List by default

v1.4.2 (2020-02-02)
- Add scroll bar to List, DropDown, Table and TreeView
- Add SetDoneFunc to TreeView
- Fix mouse support when resuming after Application.Suspend
- Fix rendering issues with TextViews which have their background color set to
ColorDefault

v1.4.1 (2020-01-23)
- Add ProgressBar widget
- Add Application.RingBell
- Add example multi-layout application and mouse-enabled application
- Add documentation for GetChildren methods

v1.4.0 (2020-01-16)
- Bump version to resolve issues with "go get"

v0.2.2 (2020-01-06)
- Add optional message displayed after CheckBox
- Fix Dropdown mouse capture behavior
- Fix TextView region highlighting on last line

v0.2.1 (2020-01-01)
- Add initial mouse support (some widgets are unsupported)
- Add window size change handler
- Fix resetting TextView color with [-]
- Fix adding unnecessary padding to some runes
- Fix handling ANSI color escape codes 39 (default foreground) and 49 (default
background)

v0.2.0 (2019-12-30)
- Fork tview as cview

v0.1.20 (2019-07-08)
- Added autocomplete functionality to InputField.

v0.1.19 (2018-10-28)
- Added QueueUpdate() and QueueEvent() to Application to help with
modifications to primitives from goroutines.

v0.1.18 (2018-10-18)
- InputField elements can now be navigated freely.

v0.1.17 (2018-06-20)
- Added TreeView.

v0.1.15 (2018-05-02)
- Flex and Grid don't clear their background per default, thus allowing for
custom modals.

v0.1.14 (2018-04-13)
- Added an Escape() function which keep strings like color or region tags from
being recognized as such.
- Added ANSIWriter() and TranslateANSI() which convert ANSI escape sequences to
tview color tags.

v0.1.13 (2018-04-01)
- Added background colors and text attributes to color tags.

v0.1.12 (2018-03-13)
- Added "suspended mode" to Application.

v0.1.11 (2018-03-02)
- Added a RemoveItem() function to Grid and Flex.

v0.1.10 (2018-02-22)

cview/ansi.go
package cview

import (
	"bytes"
	"fmt"
	"io"
	"strconv"
	"strings"
)

// The states of the ANSI escape code parser.
const (
	ansiText = iota
	ansiEscape
	ansiSubstring
	ansiControlSequence
)

// ansi is a io.Writer which translates ANSI escape codes into cview color
// tags.
type ansi struct {
	io.Writer

	// Reusable buffers.
	buffer                        *bytes.Buffer // The entire output text of one Write().
	csiParameter, csiIntermediate *bytes.Buffer // Partial CSI strings.
	attributes                    string        // The buffer's current text attributes (a tview attribute string).

	// The current state of the parser. One of the ansi constants.
	state int
}

// ANSIWriter returns an io.Writer which translates any ANSI escape codes
// written to it into cview color tags. Other escape codes don't have an effect
// and are simply removed. The translated text is written to the provided
// writer.
func ANSIWriter(writer io.Writer) io.Writer {
	return &ansi{
		Writer:          writer,
		buffer:          new(bytes.Buffer),
		csiParameter:    new(bytes.Buffer),
		csiIntermediate: new(bytes.Buffer),
		state:           ansiText,
	}
}

// Write parses the given text as a string of runes, translates ANSI escape
// codes to color tags and writes them to the output writer.
func (a *ansi) Write(text []byte) (int, error) {
	defer func() {
		a.buffer.Reset()
	}()

	for _, r := range string(text) {
		switch a.state {

		// We just entered an escape sequence.
		case ansiEscape:
			switch r {
			case '[': // Control Sequence Introducer.
				a.csiParameter.Reset()
				a.csiIntermediate.Reset()
				a.state = ansiControlSequence
			case 'c': // Reset.
				fmt.Fprint(a.buffer, "[-:-:-]")
				a.state = ansiText
			case 'P', ']', 'X', '^', '_': // Substrings and commands.
				a.state = ansiSubstring
			default: // Ignore.
				a.state = ansiText
			}

		// CSI Sequences.
		case ansiControlSequence:
			switch {
			case r >= 0x30 && r <= 0x3f: // Parameter bytes.
				if _, err := a.csiParameter.WriteRune(r); err != nil {
					return 0, err
				}
			case r >= 0x20 && r <= 0x2f: // Intermediate bytes.
				if _, err := a.csiIntermediate.WriteRune(r); err != nil {
					return 0, err
				}
			case r >= 0x40 && r <= 0x7e: // Final byte.
				switch r {
				case 'E': // Next line.
					count, _ := strconv.Atoi(a.csiParameter.String())
					if count == 0 {
						count = 1
					}
					fmt.Fprint(a.buffer, strings.Repeat("\n", count))
				case 'm': // Select Graphic Rendition.
					var background, foreground string
					params := a.csiParameter.String()
					fields := strings.Split(params, ";")
					if len(params) == 0 || len(fields) == 1 && fields[0] == "0" {
						// Reset.
						a.attributes = ""
						if _, err := a.buffer.WriteString("[-:-:-]"); err != nil {
							return 0, err
						}
						break
					}
					lookupColor := func(colorNumber int) string {
						if colorNumber < 0 || colorNumber > 15 {
							return "black"
						}
						return []string{
							"black",
							"maroon",
							"green",
							"olive",
							"navy",
							"purple",
							"teal",
							"silver",
							"gray",
							"red",
							"lime",
							"yellow",
							"blue",
							"fuchsia",
							"aqua",
							"white",
						}[colorNumber]
					}
				FieldLoop:
					for index, field := range fields {
						switch field {
						case "1", "01":
							if strings.IndexRune(a.attributes, 'b') < 0 {
								a.attributes += "b"
							}
						case "2", "02":
							if strings.IndexRune(a.attributes, 'd') < 0 {
								a.attributes += "d"
							}
						case "4", "04":
							if strings.IndexRune(a.attributes, 'u') < 0 {
								a.attributes += "u"
							}
						case "5", "05":
							if strings.IndexRune(a.attributes, 'l') < 0 {
								a.attributes += "l"
							}
						case "22":
							if i := strings.IndexRune(a.attributes, 'b'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
							if i := strings.IndexRune(a.attributes, 'd'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "24":
							if i := strings.IndexRune(a.attributes, 'u'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "25":
							if i := strings.IndexRune(a.attributes, 'l'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "30", "31", "32", "33", "34", "35", "36", "37":
							colorNumber, _ := strconv.Atoi(field)
							foreground = lookupColor(colorNumber - 30)
						case "39":
							foreground = "-"
						case "40", "41", "42", "43", "44", "45", "46", "47":
							colorNumber, _ := strconv.Atoi(field)
							background = lookupColor(colorNumber - 40)
						case "49":
							background = "-"
						case "90", "91", "92", "93", "94", "95", "96", "97":
							colorNumber, _ := strconv.Atoi(field)
							foreground = lookupColor(colorNumber - 82)
						case "100", "101", "102", "103", "104", "105", "106", "107":
							colorNumber, _ := strconv.Atoi(field)
							background = lookupColor(colorNumber - 92)
						case "38", "48":
							var color string
							if len(fields) > index+1 {
								if fields[index+1] == "5" && len(fields) > index+2 { // 8-bit colors.
									colorNumber, _ := strconv.Atoi(fields[index+2])
									if colorNumber <= 15 {
										color = lookupColor(colorNumber)
									} else if colorNumber <= 231 {
										red := (colorNumber - 16) / 36
										green := ((colorNumber - 16) / 6) % 6
										blue := (colorNumber - 16) % 6
										color = fmt.Sprintf("#%02x%02x%02x", 255*red/5, 255*green/5, 255*blue/5)
									} else if colorNumber <= 255 {
										grey := 255 * (colorNumber - 232) / 23
										color = fmt.Sprintf("#%02x%02x%02x", grey, grey, grey)
									}
								} else if fields[index+1] == "2" && len(fields) > index+4 { // 24-bit colors.
									red, _ := strconv.Atoi(fields[index+2])
									green, _ := strconv.Atoi(fields[index+3])
									blue, _ := strconv.Atoi(fields[index+4])
									color = fmt.Sprintf("#%02x%02x%02x", red, green, blue)
								}
							}
							if len(color) > 0 {
								if field == "38" {
									foreground = color
								} else {
									background = color
								}
							}
							break FieldLoop
						}
					}
					var colon string
					if len(a.attributes) > 0 {
						colon = ":"
					}
					if len(foreground) > 0 || len(background) > 0 || len(a.attributes) > 0 {
						fmt.Fprintf(a.buffer, "[%s:%s%s%s]", foreground, background, colon, a.attributes)
					}
				}
				a.state = ansiText
			default: // Undefined byte.
				a.state = ansiText // Abort CSI.
			}

			// We just entered a substring/command sequence.
		case ansiSubstring:
			if r == 27 { // Most likely the end of the substring.
				a.state = ansiEscape
			} // Ignore all other characters.

			// "ansiText" and all others.
		default:
			if r == 27 {
				// This is the start of an escape sequence.
				a.state = ansiEscape
			} else {
				// Just a regular rune. Send to buffer.
				if _, err := a.buffer.WriteRune(r); err != nil {
					return 0, err
				}
			}
		}
	}

	// Write buffer to target writer.
	n, err := a.buffer.WriteTo(a.Writer)
	if err != nil {
		return int(n), err
	}
	return len(text), nil
}

// TranslateANSI replaces ANSI escape sequences found in the provided string
// with cview's color tags and returns the resulting string.
func TranslateANSI(text string) string {
	var buffer bytes.Buffer
	writer := ANSIWriter(&buffer)
	writer.Write([]byte(text))
	return buffer.String()
}
